### PR TITLE
Move to VMC standalone

### PR DESCRIPTION
**Pull request by benedikt-voelkel**
introduce necessary recipe updates

* add to simulation recipe but make also explicit in individual VMC
  recipes

* remove ENABLE_VMC

* O2sim built tested with altered O2 and AEGIS locally

* full_system_test runs locally

* disable VMC where ROOT5 is built

* move to foreseen AEGIS branch alice/v1.4

* keep `VMC` as explicit dependency in `simulation` recipe to be more transparent

* provide symlink `libVMC.so` -> `libVMCLibrary.so`, in particular for older scripts that rely on things like `gSystem->Load("libVMC")`

### CODE REVIEW

**Comment by benedikt-voelkel** [comment]
@TimoWilken @ktf 

The CI will break at this point. Can we just stop it now already?

**Comment by TimoWilken** [comment]
I've marked the PR as draft, that'l stop the PR from running.

**Comment by benedikt-voelkel** [comment]
@ihrivnac @sawenzel 
Here is a draft PR. I could build `O2sim`

As mentioned by @davidrohr on JIRA (https://alice.its.cern.ch/jira/browse/O2-2689?filter=-3) AEGIS needs a twist as well which I did locally.

I ran the `full_system_test` which works just fine

Let's discuss details here. In the meantime I will open a PR for O2 and AEGIS.

Also, another test with `aliBuild build AliPhysics --defaults release` is running which should build with ROOT5. Let's see, I expect that to not be affected at all

**Comment by benedikt-voelkel** [comment]
As a ROOT5 build:
`aliBuild build AliPhysics --defaults release` runs through, all `AliRoot` and `AliPhysics` tests succeed

**Comment by sawenzel** [comment]
To me, this looks ok.

**Comment by benedikt-voelkel** [comment]
I am working on `AliRoot` at the moment and will check if that works as well. I am going to post the result here. If that goes through as well, I will open the O2 and AliRoot PRs

**Comment by benedikt-voelkel** [comment]
Be backwards compatible and provide a symlink `libVMC.so` to `libVMCLibrary.so` so that `AliPhysics` and `AliRoot` macros can find it when using things like `gSystem->Load("libVMC")`

File: aegis.sh
Comment (ihrivnac) [inline aegis.sh:6]: This dependence is surprising; I looked in the code and the only usage of VMC is to access mag field in GeneratorCosmics:
 ```c++
 //
  auto fld = TGeoGlobalMagField::Instance()->GetField();
  if (TVirtualMC::GetMC() && TVirtualMC::GetMC()->GetMagField()) {
    fld = TVirtualMC::GetMC()->GetMagField();
  }
```
Does it make sense? I thought, that we always define field via TGeo and then set it to VMC.

File: simulation.sh
Comment (ihrivnac) [inline simulation.sh:4]: I think that this is not needed, as the dependence on VMC is already declared in GEANT3 and GEANT4_VMC

File: vmc.sh
Comment (ihrivnac) [inline vmc.sh:22]: Why is this link needed?

Comment (ihrivnac) [reply vmc.sh:22]: I see, that you commented this above.

File: simulation.sh
Comment (benedikt-voelkel) [reply simulation.sh:4]: It is not needed, just to make it explicit. But we can remove it of course

File: vmc.sh
Comment (benedikt-voelkel) [reply vmc.sh:22]: It's when ROOT macros from AliRoot or AliPhysics do something like `gSystem->Load("libVMC")`

File: simulation.sh
Comment (benedikt-voelkel) [reply simulation.sh:4]: indeed @ihrivnac should we make it explicit in the `fairroot.sh` recipe rather than here? I would just like to have it explicitly in central recipes so that it is clear what happens.

Comment (ihrivnac) [reply simulation.sh:4]: I don't see a need for making it explicit here or in FairRoot, as the VMC is anyway nowhere else used (except for the AEGIS case where its usage is disputable)

File: aegis.sh
Comment (benedikt-voelkel) [reply aegis.sh:6]: mark as resolved for now, coming back to it if necessary

**Comment by benedikt-voelkel** [comment]
we should merge https://github.com/AliceO2Group/AEGIS/pull/11 and https://github.com/AliceO2Group/AliceO2/pull/7841 before we start the CI chain here

File: simulation.sh
Comment (ihrivnac) [reply simulation.sh:4]: ROOT dependency can be dropped in FairRoot. We have also many explicit dependencies on ROOT in O2, and we do not add ROOT in requires in o2.sh as it is already implicitly included. Do you find this misleading as well?

File: fluka_vmc.sh
Comment (ihrivnac) [inline fluka_vmc.sh:7]: We can remove ROOT as VMC already depends on ROOT

File: geant3.sh
Comment (ihrivnac) [inline geant3.sh:5]: We can remove ROOT as VMC already depends on ROOT

File: geant4_vmc.sh
Comment (ihrivnac) [inline geant4_vmc.sh:6]: We can remove ROOT as VMC already depends on ROOT

File: simulation.sh
Comment (benedikt-voelkel) [reply simulation.sh:4]: ok, true. If we remove ROOT from FairRoot then I agree
* remove ROOT from all VMC recipes
* remove ROOT from FairRoot
* remove VMC from simulation

I will test that and add it to the PR

File: geant3.sh
Comment (benedikt-voelkel) [reply geant3.sh]: removed

File: geant4_vmc.sh
Comment (benedikt-voelkel) [reply geant4_vmc.sh]: removed

File: fluka_vmc.sh
Comment (benedikt-voelkel) [reply fluka_vmc.sh]: removed

File: simulation.sh
Comment (benedikt-voelkel) [reply simulation.sh]: I have removed the requirements as listed in my previous comment

Comment (ihrivnac) [reply simulation.sh]: Great, thank you

**Comment by ktf** [comment]
Please remember to remove [Draft] and change it a normal PR when you are ready.

**Comment by davidrohr** [comment]
FairROOT seems to fail to build without ROOT

File: fluka_vmc.sh
Comment (ihrivnac) [reply fluka_vmc.sh]: It turns out that this is required when we build FairRoot without simulation; can you put ROOT back in fairroot.sh?

**Comment by benedikt-voelkel** [comment]
Hello @ktf
It seems the failing mac CI is unrelated to the changes. Seems to be a problem with `autotools` there

**Comment by ktf** [comment]
@MichaelLettrich was the OS updated?

**Comment by MichaelLettrich** [comment]
@benedikt-voelkel  @ktf : the macOS version is Big Sur (11.6.1)

```bash
alibuild@aido2osx3 ~ % xcodebuild -version
Xcode 13.1
Build version 13A1030d

alibuild@aido2osx3 ~ % which autoconf      
/usr/local/bin/autoconf

alibuild@aido2osx3 ~ % autoconf --version
autoconf (GNU Autoconf) 2.71
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+/Autoconf: GNU GPL version 3 or later
<https://gnu.org/licenses/gpl.html>, <https://gnu.org/licenses/exceptions.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by David J. MacKenzie and Akim Demaille.
```

anything else you need?

**Comment by ktf** [comment]
Maybe we should reboot the machine? @MichaelLettrich can you take care of it?

**Comment by MichaelLettrich** [comment]
@ktf done

**Comment by benedikt-voelkel** [comment]
Thanks @MichaelLettrich 
Could you by any chance restart the macOS CI? I don't seem to have the power to do so

**Comment by MichaelLettrich** [comment]
@benedikt-voelkel  I've restarted the PR and it looks like it reproduces the same error.

**Comment by benedikt-voelkel** [comment]
Thanks @MichaelLettrich 
I saw it, looks weird to me as I can't see how this is related to any of the changes I made.
Maybe it is possible to run on another mac machine? Unfortunately, I don't have access to a mac myself, in that case I would try to reproduce/see what the problem is.

**Comment by MichaelLettrich** [comment]
All alidist PRs are built on the same mac and we clearly have PRs that pass the CI, so it should be related to this PR.

**Comment by benedikt-voelkel** [comment]
With Michael's help I was able to build `aliBuild build O2Suite --defaults o2` on the mac machine `aido2osx3` in `/Users/alibuild/bvolkel`
I can run the simulation and the library path to the VMC lib is correctly resolved (see `/Users/alibuild/bvolkel/test_vmc_standalone/dyld.log`)
Is it possible that for this CI the build area is somehow corrupted?

**Comment by MichaelLettrich** [comment]
@ktf @TimoWilken  so now the CI machine was restarted, the build area was cleared and @benedikt-voelkel managed to build it by hand on the CI machine but the build still fails when running on CI. Is there anything specific in the CI environment that can trigger the issue on macOS ?

**Comment by benedikt-voelkel** [comment]
Indeed, it is the same as before. I don't know what I can do from here. As I said, when I did it on the machine, the only thing was checking out this `alidist` branch and running `aliBuild` as described above.
If anyone has a hint, please  let me know.
Again, `autotools` is built way before any of the packages I changed the recipes for, right? I don't see how my changes could have an impact on that.

I could of course rebase the PR on top of the latest `alidist@master` and trigger a new CI run if that would be a possibility.

**Comment by ktf** [comment]
In the past the autotools issues have been issues with XCode on the machine not being properly updated. I will merge this, since the arm version is green. Let's keep an eye on it.

**Comment by ktf** [comment]
Actually, before we merge this, since it will trigger a full rebuild, @TimoWilken could you please warm up the cache?

Alternatively, do we really need to modify so many defaults? Can't we change those where we need to say "NO" or have some logic which is default independent?

**Comment by ihrivnac** [comment]
On my Mac (11.5.2), autotools are taken from the system:
```
[MacBook-Pro-5] ~/work/projects/aliceO2 > aliBuild --defaults o2 -j 4 build O2

==> aliBuild can take the following packages from the system and will not build them:
      GSL, FreeType, CMake, libxml2, zlib, libpng, autotools, lzma, bz2, `lz4`
```

**Comment by benedikt-voelkel** [comment]
Concerning the defaults:
In the `VMC` recipe we could check if ROOT is version 5. In that case we consider the `VMC` build done, not doing anything. You think that would work? I can try it in that case.

That means that the `release` and `prod-latest` defaults need no change **IF** we then also put `ROOT` back to the `Geant3` and `Geant4_VMC` recipes. Because otherwise - in these `ROOT5` modes, they wouldn't find `ROOT`

I would say, all other defaults containing `ENABLE_VMC` should be changed so this logic disappears.

**Comment by ktf** [comment]
To me it sounds like a good idea. Having to rebuild the whole stack for this kind of change is a bit of a pain...

**Comment by benedikt-voelkel** [comment]
Ok, I will give that a shot! Thanks!

File: vmc.sh
Comment (benedikt-voelkel) [inline vmc.sh:39]: Is this approach acceptable?

Comment (ktf) [reply vmc.sh:39]: apart from the fact I would do:

```bash
case $ROOT_VERSION in
  6*) ... ;;
  *) .. ;;
esac
```

it seems fine to me. You could also generate the module file once, before actually building the package.

File: vmc.sh
Comment (ktf) [inline vmc.sh:17]: There is already ROOT_VERSION which is defined by alibuild the moment you require root.

File: defaults-o2.sh
Comment (ktf) [inline defaults-o2.sh:6]: Does it arm to leave it there for now? this will trigger a rebuild as well...

Comment (benedikt-voelkel) [reply defaults-o2.sh:6]: ok, will leave that and only remove the corresponding line in `root.sh`

File: vmc.sh
Comment (benedikt-voelkel) [reply vmc.sh]: I used that as suggested

File: defaults-o2.sh
Comment (benedikt-voelkel) [reply defaults-o2.sh]: no defaults are changed now

File: vmc.sh
Comment (benedikt-voelkel) [reply vmc.sh]: Modulfile is created at the beginning and only updated with the corresponding env variables if VMC is actually built

**Review by ktf — APPROVED** [review]

**Comment by benedikt-voelkel** [comment]
Concerning failing mac CI this time: https://github.com/AliceO2Group/O2Physics/pull/432#issuecomment-1012408319

**Comment by ktf** [comment]
The error is indeed genuine, albeit unrelated. I would say we can merge this after @TimoWilken has warmed up the cache.

**Comment by TimoWilken** [comment]
Caching is done!

**Pull request merged**
